File: sol/solution/0001-0100/0039. Combination Sum/Solution.swift
class Solution {
    func combinationSum(_ candidates: [Int], _ target: Int) -> [[Int]] {
        var result = [[Int]]()
        var combination = [Int]()
        let sortedCandidates = candidates.sorted()
        func backtrack(_ start: Int, _ remaining: Int) {
            if remaining == 0 {
                result.append(combination)
                return
            }
            for i in start..<sortedCandidates.count {
                if sortedCandidates[i] > remaining {
                    break
                }
                combination.append(sortedCandidates[i])
                backtrack(i, remaining - sortedCandidates[i])
                combination.removeLast()
            }
        }
        backtrack(0, target)
        return result
    }
}

File: sol/solution/0001-0100/0039. Combination Sum/Solution.dart
class Solution {
  List<List<int>> combinationSum(List<int> candidates, int target) {
    List<List<int>> result = [];
    List<int> currentCombination = [];
    dfs(candidates, target, currentCombination, 0, result);
    return result;
  }
  void dfs(
    List<int> candidates,
    int target,
    List<int> currentCombination,
    int startIndex,
    List<List<int>> result,
  ) {
    if (target == 0) {
      result.add([...currentCombination]);
      return;
    }
    if (target < 0) {
      return;
    }
    for (int i = startIndex; i < candidates.length; i++) {
      currentCombination.add(candidates[i]);
      dfs(candidates, target - candidates[i], currentCombination, i, result);
      currentCombination.removeLast();
    }
  }
}